mne/preprocessing/xdawn.py

```diff
@@ -0,0 +1,566 @@
+    # than the time window.
+    return isi.min() < window
+
+
+def _construct_signal_from_epochs(epochs):
+    """Reconstruct pseudo continuous signal from epochs."""
+    start = (np.min(epochs.events[:, 0])
+             + int(epochs.tmin * epochs.info['sfreq']))
+    stop = (np.max(epochs.events[:, 0])
+            + int(epochs.tmax * epochs.info['sfreq']) + 1)
+
+    n_samples = stop - start
+    epochs_data = epochs.get_data()
+    n_epochs, n_channels, n_times = epochs_data.shape
+    ix_events = epochs.events[:, 0] - epochs.events[0, 0]
+
+    data = np.zeros((n_channels, n_samples))
+    for idx in range(n_epochs):
+        onset = ix_events[idx]
+        offset = onset + n_times
+        data[:, onset:offset] = epochs_data[idx]
+
+    return data
+
+
+def least_square_evoked(epochs, return_toeplitz=False):
+    """Least square estimation of evoked response from a Epochs instance.
+
+    Parameters
+    ----------
+    epochs : Epochs instance
+        An instance of Epochs.
+    return_toeplitz : bool
+        If true, compute the toeplitz matrix.
+
+    Returns
+    -------
+    evokeds : dict of evoked instance
+        An dict of evoked instance for each event type in epochs.event_id.
+    toeplitz : dict of array
+        If return_toeplitz is true, return the toeplitz matrix for each event
+        type in epochs.event_id.
+    """
+    if not isinstance(epochs, _BaseEpochs):
+        raise ValueError('epochs must be an instance of `mne.Epochs`')
+
+    events = epochs.events.copy()
+    events[:, 0] -= events[0, 0] + int(epochs.tmin * epochs.info['sfreq'])
+    data = _construct_signal_from_epochs(epochs)
+    evoked_data, toeplitz = _least_square_evoked(data, events, epochs.event_id,
+                                                 tmin=epochs.tmin,
+                                                 tmax=epochs.tmax,
+                                                 sfreq=epochs.info['sfreq'])
+    evokeds = dict()
+    info = cp.deepcopy(epochs.info)
+    for name, data in evoked_data.items():
+        n_events = len(events[events[:, 2] == epochs.event_id[name]])
+        evoked = EvokedArray(data, info, tmin=epochs.tmin,
+                             comment=name, nave=n_events)
+        evokeds[name] = evoked
+
+    if return_toeplitz:
+        return evokeds, toeplitz
+
+    return evokeds
+
+
+class Xdawn(TransformerMixin, ContainsMixin):
+
+    """Implementation of the Xdawn Algorithm.
+
+    Xdawn is a spatial filtering method designed to improve the signal
+    to signal + noise ratio (SSNR) of the ERP responses. Xdawn was originaly
+    designed for P300 evoked potential by enhancing the target response with
+    respect to the non-target response. This implementation is a generalization
+    to any type of ERP.
+
+    Parameters
+    ----------
+    n_components : int, default 2
+        The number of components to decompose M/EEG signals.
+    signal_cov : None, Covariance instance or ndarray
```

Comment:
None | Covariance | ndarray of float, shape(n_channels, n_channels)

-- in general take a look at the other doc strings and adopt